Second pass on hacking.rst

stasm opened the pull request:
I applied the feedback we got on the first iteration of the Hacking documentation.

@Pike, @zbraniecki — r?

zbraniecki commented on docs/hacking.rst:
```diff
@@ -2,273 +2,256 @@
-Each layer has different external dependencies and is intended to tie better into the
-host and client environments.
+:ECMA 402 Intl
+    The low-level API providing L20n's primitives and building-blocks.
```
So, I'm concerned about using the term "ECMA" here as misleading. Some of those building blocks are not yet in ECMA and I'd prefer not to pretend that this is ECMA.
In particular the most important class - MessageContext - has not been even suggested to ECMA yet.

Pike replied:
Similarily, I think we should call this something like "L20n builds on top of JavaScript Intl APIs. We also work with ECMA to enhance these APIs where there's overlap between the needs of L20n and other libraries needing access to internationalized data."

zbraniecki commented on docs/hacking.rst:
```diff
@@ -2,273 +2,256 @@
+At the heart of l20n.js is the low-level ECMA 402's Intl API.  L20n.js uses 
+``Intl.NumberFormat`` to format numbers in a locale-aware manner and 
+``Intl.DateTimeFormat`` to format dates.  Other features are provided by our custom
+extensions to the Intl API, modeled after the existing ECMA specifications:
```
Instead, I'd suggest to put here a note that states that we intend to work with ECMA on making all APIs on this layer be aligned with ECMA402.

zbraniecki commented on docs/hacking.rst:
```diff
@@ -2,273 +2,256 @@
-A single localization unit in L20n vocabulary is called an entity.
-An entity is an ``Object`` with three fields:
+A single localization unit in ``MessageContext`` is called an entity.  An entity 
+has an identifier, a value and traits.
```
Not sure if it's something to fix but you mention MessageContext before you define it.

stasm replied:
I think that's okay because ``MessageContext`` is already mentioned in the list above.

zbraniecki commented on docs/hacking.rst:
```diff
@@ -2,273 +2,256 @@
+The translations are considered untrusted. ``MessageContext`` has been 
+designed to be resilient to formatting errors.  If a translation contains 
+errors, ``MessageContext`` will try to salvage as much of the translation as 
+possible and will still return a string.
```
This reads a bit weird. The `The translations are considered untrusted` being the first sentence of the paragraph feels odd.
Maybe put it further down in the paragraph or explain a bit: `L20n takes an approach of assuming that localization resources do not go through a security review process and may be coming from a third-party, thus the translation have to be considered untrusted`.

zbraniecki commented on docs/hacking.rst:
```diff
@@ -2,273 +2,256 @@
+For performance reasons formatting errors aren't returned together with the 
+formatted string.  The ``format`` method of ``MessageContext`` takes an 
+additional parameter, ``errors``, which if provided must be an Array to which 
+errors will be pushed.
```
The first sentence sounds like returning error with the formatted string is the default programming pattern and we need to justify not following it - feels defensive :)
I believe you can do without this sentence all together.

zbraniecki commented on docs/hacking.rst:
```diff
@@ -2,273 +2,256 @@
-    Language tag associated with the bundle
+The ``Localization`` class can be thought of as a multi-lingual, asynchronous 
+wrapper on top of multiple ``MessageContext`` instances which handles lazy-IO 
+and the fallback strategy.
```
`language fallback` - I believe that it's worth pointing out what is fallbacking

Pike commented on docs/hacking.rst:
```diff
@@ -2,273 +2,256 @@
+provide different ways to build bundles, use different IO and language 
+negotiation methods, as well as pass custom functions to the ``MessageContext``
+constructor.  The ``Localization`` class itself doesn't know anything about the
+environment it runs in and doesn't provide any special methods to localize it. 
```
"environment" and "localize it" are a tad underdefined. Maybe this is easier if you start with the sentence below right away, somewhat like:

The ``LocalizationObserver`` adapts the Localization to APIs that are properly suited for programming environment.

or so?

stasm replied:
I think I'll just mention that `Localization` is env-agnostic.  The first sentence in the next section explains this better anyways.

Pike commented on docs/hacking.rst:
```diff
@@ -2,273 +2,256 @@
 Hacking
 =======
 
```
Maybe add a quick paragraph on top to clarify the target audience for this doc?

Pike commented on docs/hacking.rst:
```diff
@@ -72,21 +79,23 @@ MessageContext
-environmental variables, user-provided arguments, other messages in the current 
-context and any other syntax that the format allows.
+Translations must be defined in the FTL syntax.  FTL has been designed as an 
+extension to the `ICU MessageFormat`_ standard.
```
I'd drop the ICU stuff here. But add a link to the syntax doc instead?

stasm commented:
Thanks to you both for the comments. I addressed all feedback.  Pike already lgtm'ed this on IRC. @zbraniecki, would you like to give it another read?

zbraniecki commented:
r+
